### DISCUSSION TITLE
Sharing CWMS Computational Processor (CCP) Algorithms

### DISCUSSION THREAD

**Discussion by JasonDSS** [opendcs/opendcs#306] [Q&A] [answered]
Hello Everyone,

I'm just getting started with CCP and was wondering if anyone has any algorithms (other than the ones that already come with OpenDCS 7.0.8) that they'd be willing to share. Some example algorithms I'm looking for:

1) Computing incremental precip: I tried the one that comes with OpenDCS, and it works fine for computing hourly incremental precip from 1Hour or 15minute cumulative precip data. But it doesn't seem to work when computing hourly incremental precip from ~1Hour cumulative precip data (the ~1Hour data we have typically does not have data at the top of the hour). I suspect the algorithm needs data at the top of each hour to work? Perhaps this would require a step to interpolate the data on the hour before using this script? Which brings me to the next item:

2) Python Algorithms: Since I'm not a programmer, maybe the Python algorithm is easier to pick up for creating custom algorithms? Anyone have any examples or documentation that I can use for reference? There's a couple of examples in the OpenDCS docs, but was hoping for more examples and a full list of functions that can be used.

3) Or any algorithm that does not come with OpenDCS but may be useful for reservoir comps, such as computing outflow, storage, inflow, etc.

Thanks in advance!

Jason

**Comment by MikeNeilson** [comment, marked as answer]
1. I wouldn't expect it too, expect for specific algorithms the expectation is regular interval input. From I remember discussing previously about your data my suggestion would be to use the resample algorithm to convert everything to regular interval 15-minute (or shorter if required) data and all the computations be performed on those resulting timeseries. (yes this does add a lot of time series to your database, but you're study people and real-time modellers will thank you for having regular interval data available to use in models.

2. Generally, yes. We need more examples in the project though. @zamusacelre or @jbatterman, do you have some examples you can share with Jason?

3. There are algorithms for everything but inflow (well, depending on your particular inflow needs), and maybe outflow. There is the CwmsRatingSingleIndep which is good for stage->flow, elev->area, elev->stor ,etc if you're tables are stored in CWMS. There's also a simple Rdb rating lookup function and tabular rating function.

**Reply by JasonDSS** [reply]
Thanks Mike. I've considered switching everything over to regular interval, but it's a lot of timeseries IDs to switch over, and a lot of our stuff are already hard-coded to point to the psuedo-regular interval datasets, even the CWMS models. Regarding resampling to 15 minutes and then doing computations on the resulting timeseries, is there anyway to create intermediate datasets (ie, the 15-minute dataset from resampling) without it writing to the CWMS DB? For example, it would resample the ~1Hour data into 15-minute regular interval as an intermediate/temporary dataset. Then another computation would compute the hourly incremental precip using the intermediate/temporary dataset and store to the CWMS DB. Would this only be doable using a Python Algorithm or creating an entirely custom-made algorithm?

**Reply by MikeNeilson** [reply]
The actual conversion work could be relatively small by using Timeseries groups and a group computation. You'd want a handful of computations running on CompProc to spread the load but it's not too bad.


As for intermediary computations, no. That's not possible at this time, nor do I have a sense of how that would even work.

It's possible we could modify some of the functions to handle irregular data but that would likely be on a case by case basis. Incremental precip being fairly simple, inflow being not impossible but a royal pain.

Getting data to regular intervals probably is the better long term choice; everything is simpler.

**Reply by JasonDSS** [reply]
Are there more CCP functions in the Python algorithm than what's shown here or are these all of them?
![image](https://github.com/opendcs/opendcs/assets/143829980/560a97c4-ced8-4cfe-8d46-a4663b42ea7a)

**Reply by MikeNeilson** [reply]
All of what works in Jython is available (but for the love of sanity please don't try to do anything with DSS).

The above functions are just specific OpenDCS helpers. ... and wait, yes it appears a few are missing.

As you can see in the example there is a setOutput function, the is required to be called or you will get no data saved.

**Reply by MikeNeilson** [reply]
Thinking about it, making modifications to support irregular input is likely a good idea. but to do so generically for everything may be a bit of an undertaking. We'll at least keep it in mind.

**Reply by JasonDSS** [reply]
Is there a resample function in the Python algorithm? My guess is no, because it's not really a time-slice type of function (ie requires data points from other time slices)? But I thought I'd ask in case there was, because I'm thinking it would be really useful for us.

**Reply by MikeNeilson** [reply]
No, given the nature of the way the computation process works it wouldn't make sense at that level. 

A time slice algorithm operates on just that, an individual time slice, or sample. 
A aggregate algorithm still operates on individual samples but they are group by the given window (e.g. 1 day).
